watertap/tools/parameter_sweep/differential_parameter_sweep.py

```diff
@@ -0,0 +1,390 @@
+            ):
+                global_values[:, i] = item["value"]
+
+        self.comm.Bcast(global_values, root=0)
+
+        return global_values
+
+    def _aggregate_results(self, local_output_dict):
+
+        num_local_samples = len(local_output_dict["solve_successful"])
+
+        # Create the global results dictionary
+        global_results_dict = self._create_global_output(local_output_dict)
+
+        # Broadcast the number of global samples to all ranks
+        num_global_samples = len(global_results_dict["solve_successful"])
+        num_global_samples = self.comm.bcast(num_global_samples, root=0)
+
+        global_results_arr = self._aggregate_results_arr(
+            global_results_dict, num_global_samples
+        )
+        global_input_values = self._aggregate_input_arr(
+            global_results_dict, num_global_samples
+        )
+
+        return (
+            global_results_dict,
+            global_results_arr,
+            global_input_values,
+            num_global_samples,
+        )
+
+    def _do_param_sweep(
+        self,
+        model,
+        sweep_params,
+        differential_sweep_specs,
+        outputs,
+        local_values,
+    ):
+
+        # Initialize space to hold results
+        local_num_cases = np.shape(local_values)[0]
+
+        # Create the output skeleton for storing detailed data
+        local_output_dict = self._create_local_output_skeleton(
+            model, sweep_params, outputs, local_num_cases
+        )
+
+        local_solve_successful_list = []
+
+        if self.config.reinitialize_function is not None:
+            reinitialize_values = ComponentMap()
+            for v in model.component_data_objects(pyo.Var):
+                reinitialize_values[v] = v.value
+        else:
+            reinitialize_values = None
+
+        # ================================================================
+        # Run all optimization cases
+        # ================================================================
+        counter = 0
+        differential_sweep_output_dict = {}
+
+        for k in range(local_num_cases):
+
+            # Step 1 : Run baseline/nominal case
+            # Update the model values with a single combination from the parameter space
+            self._update_model_values(model, sweep_params, local_values[k, :])
+
+            if self.config.probe_function is None:
+                run_successful = self._param_sweep_kernel(
+                    model,
+                    reinitialize_values,
+                )
+            else:
+                run_successful = False
+
+            # Update the loop based on the reinitialization for baseline values
+            self._update_local_output_dict(
+                model,
+                sweep_params,
+                k,
+                local_values[k, :],
+                run_successful,
+                local_output_dict,
+            )
+
+            local_solve_successful_list.append(run_successful)
```

Comment:
This code is identical to the base class.

This case is a bit more complicated, but perhaps the easiest thing to do is have the base class `_do_param_sweep` call an optional method or callback inside its `for` loop with this class implements for "Step 2"

Reply:
Repurposed the base class method to handle this. Although, not sure if we should add assertions to catch incorrect keyword argument specifications for `ParameterSweep` and `RecursiveParameterSweep` for the config option `run_differential_sweep`. This may require us to write `__init__` functions for the two as, ideally, we would like to catch that when initializing the class. That being said, we could also check for that assertion within the `parameter_sweep` method.